presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -253,40 +239,52 @@ public void finish()
     @Override
     public boolean isFinished()
     {
-        return finishing && aggregationBuilder == null && (outputIterator == null || !outputIterator.hasNext());
+        return finishing && hashAggregationBuilder == null && (outputIterator == null || !outputIterator.hasNext());
     }
 
     @Override
     public boolean needsInput()
     {
-        return !finishing && outputIterator == null && (aggregationBuilder == null || !aggregationBuilder.checkFullAndUpdateMemory());
+        return !finishing && outputIterator == null && (hashAggregationBuilder == null || !hashAggregationBuilder.checkFullAndUpdateMemory());
     }
 
     @Override
     public void addInput(Page page)
     {
         checkState(!finishing, "Operator is already finishing");
         requireNonNull(page, "page is null");
-        if (aggregationBuilder == null) {
-            aggregationBuilder = new GroupByHashAggregationBuilder(
-                    accumulatorFactories,
-                    step,
-                    expectedGroups,
-                    groupByTypes,
-                    groupByChannels,
-                    hashChannel,
-                    operatorContext,
-                    maxEntriesBeforeSpill,
-                    memoryLimitBeforeSpill,
-                    spillerFactory);
-
-            closer.register(aggregationBuilder);
+        if (hashAggregationBuilder == null) {
+            if (step.isOutputPartial() || (maxEntriesBeforeSpill == 0 && memoryLimitBeforeSpill.toBytes() == 0)) {
```

Comment:
should there be OR instead in:
`maxEntriesBeforeSpill == 0 && memoryLimitBeforeSpill.toBytes() == 0`

if either is set to zero, we should always use in memory builder, right?